### DISCUSSION TITLE
downloaded from information file

### DISCUSSION THREAD

**Discussion by hboetes** [nicotine-plus/nicotine-plus#1849] [Ideas]
I just checked all my files for corruption and quite a few files were having defects. Of course I'm working on fixing those defects, and I'm going to test all incoming files for defects before adding them to my collection.

Of course it would be nice to inform the source about the problems with their files. To do that the source information must be stored somehow.

In a downloaded a folder a small text file could be dropped like `print "downloaded from $source" > downloaded_from.txt`

For loose files: `print "downloaded from $source" > target_file +  '.txt'`

I think this could be in the form of a plugin, for people who'd like to see that information.

**Comment by slook** [comment]
I don't get how this is different from the `transfers.log` ?

What sort of source information are you using to detect defects, are you referring to the reported size vs. the actual size?

> it would be nice to inform the source about the problems with their files

It would be better to avoid defects in the first place. I suspect that problems are sometimes created when a file is being transferred at the same time it is modified.

**Reply by hboetes** [reply]
> I don't get how this is different from the `transfers.log` ?

Right, transfers.log is what I need.

> What sort of source information are you using to detect defects, are you referring to the reported size vs. the actual size?
I'm talking about defects that existed before the file transfer.

I found them with `flac -wst file.flac`

**Reply by mathiascode** [reply]
> It would be better to avoid defects in the first place. I suspect that problems are sometimes created when a file is being transferred at the same time it is modified.

The question is if such issues stem from Nicotine+ or the official clients. There have been several discussions about corrupted downloads with the official client over the years, and I'm not aware of any issues on our end, but I could of course be wrong. We check the offset of the incomplete file before resuming a download, instead of relying on whatever offset Nicotine+ remembers, so the room for error here should be fairly small.

**Reply by hboetes** [reply]
I bet those problems were pre-existing. Simple bit-rot being copied again and again.

**Reply by hboetes** [reply]
It would be nice if there was a way to automatically test and report bit-rot on downloaded files. Also pleasant to give the uploader a heads-up about it.

**Reply by slook** [reply]
The quality of data seems to be something that is becoming more and more important to the community as the expectations for high-res audio nowadays. Unfortunately much pollution probably comes from uneducated folk who transcode between formats, for example wrapping data from a 128kbps MP3 into a FLAC, which can be detected by [doing a spectrum analysis after the download is finished](https://github.com/nicotine-plus/nicotine-plus/issues/1847), although the only way to be absolutely sure is to do a manual comparison between multiple copies, considering that it is very likely that source audio data is obtained from unreliable sources before it appears on slsk, or on any network for that matter.

In my experience often the lack of proper artwork image or other ID3 tags such as the presence of an ISRC gives a good clue as to if the file is not the "original" copy, it is unfortunate that relevant tags are not visible in the remote user's database of shared files until each file is downloaded (also, the ability to search by ISRC would be a massive boon).

Many users use tools such as cddb or Musicbrainz to establish if files are "genuine" copies or not, but unfortunately these same tools can be used to "correct" missing tag data which further impacts the problem of data pollution when these "corrected" files are re-shared, when it would infact be better for everybody if dodgy files were simply deleted.

> We check the offset of the incomplete file before resuming a download

It would be unlikely that bit errors arise during transfers, although not out of the question as you say. For example it might be possible for the contents of a file to change without it's size being different, a way to detect this would be a CRC, comparing the MD5 hashes after the transfer is finished, it is a shame that there is no mechanism for this built into the slsk protocol such as there is for something like the bittorrent protocol.

> It would be nice if there was a way to automatically test and report bit-rot on downloaded files.

This would be achievable in a plugin, by using `download_finished_notification()` to run a test using the external `flac` command and then somehow gathering the warning output from stdout to decide if a private message needs to be sent to the user. It's a nice idea.

**Reply by hboetes** [reply]
> > It would be nice if there was a way to automatically test and report bit-rot on downloaded files.
> 
> This would be achievable in a plugin, by using `download_finished_notification()` to run a test using the external `flac` command and then somehow gathering the warning output from stdout to decide if a private message needs to be sent to the user. It's a nice idea.

I think the best way is to write a script/program that can test all sorts of file extensions and simply call that program via said plugin. Then there is the question of what to do with that information.

- Send message to owner?
- Create report file somewhere?

First step, get a working script to test all sorts of extensions.

**Reply by slook** [reply]
> the best way is to write a script/program ... and simply call that program via said plugin

I wouldn't fancy writing something from scratch, perhaps you are referring to a script which calls an already existing external program, whilst this could be easier to initially develop, it would be difficult to make it cross-platform, as you would need to do a bash script and a vbs script or whatever. Whilst calling external programs in Python isn't always straight-forward, it is possible.

Take for instance a program like `mpck` Checkmate MP3 Checker is a free program that checks MP3 files for errors (I've not tried this btw)...
https://github.com/Sjord/checkmate
http://checkmate.gissen.nl/manual.php

According to it's documentation, `mpck` returns the standard `0` or `1` depending upon if the file is "Ok" or "Bad". I don't know how to capture the return code in Python but I'm sure it's obvious pretty easily, without the need for an external script to handle this. Capturing stdout is more difficult but not impossible. Infact that's probably quite difficult in bash and perhaps easier to do in Python.

Also perhaps worth a look a Python program that can be run without a front-end (again I have not tried this):
https://sourceforge.net/projects/mp3val/

As for other extensions, FLAC, WAV, idk what factors need testing for, but it always seems to be based around frames being the correct bit position and what have you. I think tinytag has some functionality around this, idk if that might be useful or not.

I know that `ffmpeg` and `flac` programs are present on many systems and provide functions for testing of various parameters.

WAV's are more difficult since uncompressed PCM encoding is a linear bit-by-bit type of job, so each sample is simply assumed to be correct, but maybe it can be checked for correct headers at least.

WMA's are the easiest to process, since what is always needed is to delete the file, ideally without any prompt or warning!

Another possible implementation might be to perform an integrity check on local shared files, perhaps during or after building/appending the Nicotine+ shared file database handled by Scanner.

> Then there is the question of what to do with that information.

In the case of files being downloaded from a remote user, a report file serves little value unless it is used to send a private message to the user. Perhaps renaming the file with "bad" or moving it into a "bad" folder would be helpful.

In the case of local file(s) with error(s) a report would be useful, ideally with the ability to automatically exclude any bad files from the local shares database in order to help reduce pollution on the network in the first place.
